bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -1201,6 +1226,36 @@ public ByteBuf getExplicitLac(long ledgerId) throws IOException, Bookie.NoLedger
         return lac;
     }
 
+    /**
+     * Force sync given 'ledgerId' entries on the journal to the disk.
+     * This is useful for ledgers with DEFERRED_SYNC write flag
+     * for which the client receives acknoledgement for writes
+     * without the guarantee that data has been sync'd to disk (ackBeforeSync=true).
+     */
+    public void forceLedger(long ledgerId, WriteCallback cb,
+                            Object ctx, byte[] masterKey)
+                throws IOException, BookieException {
+        long requestNanos = MathUtils.nowInNano();
+        boolean success = false;
+        try {
+            LedgerDescriptor handle = handles.getHandle(ledgerId, masterKey);
```

Comment:
why do you need get handle here?

Reply:
Two (weak) reasons:
- validate masterkey
- execute inside the synchronized block like addEntry code path.

I have dropped the test about isFenced so the synchronized block maybe is useless.
All we are doing is to write on journal.

Code inside ensureLedgerOnMasterKeyCache should be treated as a per-ledger critical section ad wr are writing the meta entry the first time are seeing the ledger on the bookie.

Overall I think it makes sense and make forceLedger very similar to addEntry, but if you think it is an error or it could be a bottleneck I can drop it

Reply:
As discussed in slack I will drop it.
There is no need to touch ledger storage for force Ledger.
I will update the patch soon